include/seastar/core/shared_mutex.hh

```diff
@@ -234,6 +237,46 @@ with_lock(shared_mutex& sm, Func&& func) noexcept {
     }
 }
 
+/// \brief Construct a RAII-based shared lock corresponding to a given Mutex object.
+///
+/// Since constructors cannot be exectued asynchronously, use this function
+/// to shared lock a Mutex and construct an std::shared_lock object
+/// that will unlock the mutex when it's being destroyed.
+///
+/// The caller is responsible for keeping the corresponding Mutex object
+/// alive at least until the returned lock is destroyed.
+template <typename Mutex>
+future<std::shared_lock<Mutex>> get_shared_lock(Mutex& mutex) {
+    co_await mutex.lock_shared();
+
+    try {
+        co_return std::shared_lock<Mutex>{mutex, std::adopt_lock_t{}};
+    } catch (...) {
```

Comment:
Do you have an idea how the std::shared_lock{..., adopt_lock} could possibly throw? Maybe it's a bug in the standard that they forgot the "noexcept" for this constructor?

Reply:
It surprised me too and I don't see how it would happen either. I think it's still better to abide by the standard in case there are reasons for that. I highly doubt it's a mistake considering the type has been part of the language for 10 years

Reply:
The only guess I have is that because adopt_lock_t has undefined behavior if called on an _unlocked_ mutex, maybe they wanted to allow the possibility that this behavior might be throwing an exception...
In any case, what you did to work around this problem seems fine, so I'm resolving this discussion.